File: flashmind-back/src/main/java/com/flashmind/model/Deck.java
package com.flashmind.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "decks")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Deck {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 120)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(length = 20)
    @Builder.Default
    private String icon = "📚";

    @Column(length = 30)
    @Builder.Default
    private String color = "#7C3AED";

    @Column(name = "is_public", nullable = false)
    @Builder.Default
    private Boolean isPublic = false;

    @Column(name = "allow_clone", nullable = false)
    @Builder.Default
    private Boolean allowClone = true;

    @OneToMany(mappedBy = "deck", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Card> cards = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    @Column(name = "updated_at", nullable = false)
    @Builder.Default
    private LocalDateTime updatedAt = LocalDateTime.now();

    @PreUpdate
    public void preUpdate() { this.updatedAt = LocalDateTime.now(); }
}

File: flashmind-front/models/Deck.js
import { Api } from '../utils/api.js'

export class Deck {
  constructor(data) {
    this.id          = data.id
    this.user_id     = data.userId
    this.name        = data.name
    this.description = data.description || ''
    this.icon        = data.icon || '📚'
    this.color       = data.color || '#7C3AED'
    this.is_public   = data.isPublic ?? false
    this.allow_clone = data.allowClone ?? true
    this.card_count  = data.cardCount ?? 0
    this.due_count   = data.dueCount ?? 0
    this.created_at  = data.createdAt
    this.updated_at  = data.updatedAt
  }

  static async fetchAll() {
    const data = await Api.get('/decks')
    return data.map(d => new Deck(d))
  }

  static async fetchById(id) {
    const data = await Api.get(`/decks/${id}`)
    return new Deck(data)
  }

  static async create(payload) {
    const data = await Api.post('/decks', payload)
    return new Deck(data)
  }

  async update(payload) {
    const data = await Api.put(`/decks/${this.id}`, payload)
    return new Deck(data)
  }

  async delete() {
    return Api.delete(`/decks/${this.id}`)
  }

  async share() {
    return Api.post(`/decks/${this.id}/share`, {})
  }

  async updateShareVisibility(isPublic) {
    return Api.patch(`/decks/${this.id}/share/visibility`, { isPublic })
  }
}
